drivers/kscan/kscan_cap1203.c

```diff
@@ -0,0 +1,267 @@
+
+#define REG_MAIN_CONTROL 0x0
+#define CONTROL_INT 0x1
+
+#define REG_INPUT_STATUS 0x03
+
+#define REG_INTERRUPT_ENABLE 0x27
+#define INTERRUPT_ENABLE 0x7
+#define INTERRUPT_DISABLE 0x0
+
+struct cap1203_config {
+	struct i2c_dt_spec i2c;
+	struct gpio_dt_spec int_gpio;
+};
+
+struct cap1203_data {
+	struct device *dev;
+	kscan_callback_t callback;
+	struct k_work work;
+	/* Interrupt GPIO callback. */
+	struct gpio_callback int_gpio_cb;
+#ifdef CONFIG_KSCAN_CAP1203_POLL
+	/* Timer (polling mode). */
+	struct k_timer timer;
+#endif
+};
+
+static int cap1203_clear_interrupt(const struct i2c_dt_spec *i2c)
+{
+	uint8_t ctrl;
+	int r;
+
+	r = i2c_reg_read_byte_dt(i2c, REG_MAIN_CONTROL, &ctrl);
+	if (r < 0) {
+		return r;
+	}
+
+	ctrl = ctrl & ~CONTROL_INT;
+	return i2c_reg_write_byte_dt(i2c, REG_MAIN_CONTROL, ctrl);
+}
+
+static int cap1203_enable_interrupt(const struct i2c_dt_spec *i2c, uint8_t enable)
+{
+	return i2c_reg_write_byte_dt(i2c, REG_INTERRUPT_ENABLE, enable);
+}
+
+static int cap1203_process(const struct device *dev)
+{
+	const struct cap1203_config *config = dev->config;
+	struct cap1203_data *data = dev->data;
+	int r;
+	uint8_t input;
+	uint16_t col;
+	bool pressed;
+
+	r = i2c_reg_read_byte_dt(&config->i2c, REG_INPUT_STATUS, &input);
+	if (r < 0) {
+		return r;
+	}
+
+	pressed = !!input;
+	if (input & BIT(0)) {
+		col = 0;
+	}
+	if (input & BIT(1)) {
+		col = 1;
+	}
+	if (input & BIT(2)) {
+		col = 2;
+	}
+
+	LOG_DBG("event: input: %d\n", input);
+
+	/*
+	 * Clear INT bit to clear SENSOR INPUT STATUS bits.
+	 * Note that this is also required in polling mode.
+	 */
+	r = cap1203_clear_interrupt(&config->i2c);
+	if (r < 0) {
+		return r;
+	}
+
+	data->callback(dev, 0, col, pressed);
+
+	return 0;
+}
+
+static void cap1203_work_handler(struct k_work *work)
+{
+	struct cap1203_data *data = CONTAINER_OF(work, struct cap1203_data, work);
+
+	cap1203_process(data->dev);
+}
+
+static void cap1203_isr_handler(const struct device *dev,
+				struct gpio_callback *cb, uint32_t pins)
+{
+	struct cap1203_data *data = CONTAINER_OF(cb, struct cap1203_data, int_gpio_cb);
+
+	k_work_submit(&data->work);
+}
+
+#ifdef CONFIG_KSCAN_CAP1203_POLL
+static void cap1203_timer_handler(struct k_timer *timer)
+{
+	struct cap1203_data *data = CONTAINER_OF(timer, struct cap1203_data, timer);
+
+	k_work_submit(&data->work);
+}
+#endif
+
+static int cap1203_configure(const struct device *dev,
+			     kscan_callback_t callback)
+{
+	struct cap1203_data *data = dev->data;
+	const struct cap1203_config *config = dev->config;
+	int r;
+
+	if (!callback) {
+		LOG_ERR("Invalid callback (NULL)");
+		return -EINVAL;
+	}
+
+	data->callback = callback;
+
+	if (config->int_gpio.port != NULL) {
+		/* Clear pending interrupt */
+		r = cap1203_clear_interrupt(&config->i2c);
+		if (r < 0) {
+			return r;
+		}
+
+		r = cap1203_enable_interrupt(&config->i2c, INTERRUPT_ENABLE);
+	} else {
+		r = cap1203_enable_interrupt(&config->i2c, INTERRUPT_DISABLE);
+	}
```

Comment:
interrupt can be disabled once at init time, since int_gpio will be NULL all the time. why doing it every time?

Reply:
interrupt disabling doesn't have to do it every time. I'll change to do it at init time.